File: geo_3x3.rb

module Geo3x3
  module_function
  def encode(lat, lng, level)
    raise ArgumentError if level < 1
    lat = lat.to_f
    lng = lng.to_f

    if lng >= 0
      result = "E"
    else
      result = "W"
      lng += 180
    end

    lat = 90 - lat

    unit = 180.0

    (level - 1).times do
      unit /= 3
      x = (lng / unit).to_i
      y = (lat / unit).to_i
      result << "#{x + y * 3 + 1}"
      lng -= x * unit
      lat -= y * unit 
    end

    result
  end

  def encode_int(lat, lng, level)
    code = encode(lat, lng, level)
    raise ArgumentError unless code

    result = code[1..-1].to_i
    result = -result if code[0] == "W"

    result
  end

  def decode(code)
    if code.is_a? Integer
      code = code < 0 ? "W#{-code}" : "E#{code}"
    elsif code.is_a? String
      code = nil if code.length == 0
    else
      code = nil
    end

    raise ArgumentError unless code

    c = code[0]
    flg = true if c == "-" or c == "W"
    code = code[1..-1] if flg or c == "+" or c == "E"

    unit = 180.0
    lat = 0.0
    lng = 0.0
    level = 1

    code.each_char do |c|
      n = c.to_i
      break if n == 0

      unit = unit / 3
      n -= 1
      lng += n % 3 * unit
      lat += n / 3 * unit
      level += 1
    end

    lat += unit / 2
    lng += unit / 2
    lat = 90 - lat
    lng -= 180.0 if flg

    [lat, lng, level, unit]
  end

  def coords(code)
    pos = decode(code)
    x = pos[0]
    y = pos[1]
    u2 = pos[3] / 2
    [
      y - u2, x - u2,
      y - u2, x + u2,
      y + u2, x + u2,
      y + u2, x - u2
    ]
  end
end

File: Geo3x3.m
#import <Foundation/Foundation.h>
#import "Geo3x3.objc.h"

@implementation Geo3x3
+ (NSString*)Encode:(double)lat lng:(double)lng level:(int)level {
    if (level < 1) {
        return @"";
    }
    NSString* res = nil;
    if (lng >= 0) {
        res = @"E";
    } else {
        res = @"W";
        lng += 180.0;
    }
    lat += 90.0;
    double unit = 180.0;
    for (int i = 1; i < level; i++) {
        unit /= 3.0;
        const int x = (int)(lng / unit);
        const int y = (int)(lat / unit);
        const int n = x + y * 3 + 1;
        NSString* c = [[NSNumber numberWithInt:n] stringValue];
        res = [res stringByAppendingString:c];
        lng -= x * unit;
        lat -= y * unit;
    }
    return res;
}
+ (NSArray*)Decode:(NSString*)code {
    const int clen = [code length];
    if (code == nil || clen == 0) {
        return nil;
    }
    NSString* c = [code substringWithRange:NSMakeRange(0, 1)];
    bool flg = false;
    int begin = 0;
    if ([c isEqualToString:@"W"] || [c isEqualToString:@"-"]) {
        flg = true;
        begin = 1;
    } else if ([c isEqualToString:@"E"] || [c isEqualToString:@"+"]) {
        begin = 1;
    }
    double unit = 180.0;
    double lat = 0.0;
    double lng = 0.0;
    int level = 1;
    for (int i = begin; i < clen; i++) {
        NSString* c = [code substringWithRange:NSMakeRange(i, 1)];
        const NSRange range = [@"123456789" rangeOfString:c];
        if (range.location == NSNotFound) {
            break;
        }
        const int n = range.location;
        unit /= 3;
        lng += (double)(n % 3) * unit;
        lat += (double)(n / 3) * unit;
        level++;
    }
    lat += unit / 2;
    lng += unit / 2;
    lat -= 90.0;
    if (flg) {
        lng -= 180.0;
    }
    return [NSArray arrayWithObjects:
        [NSNumber numberWithFloat:lat],
        [NSNumber numberWithFloat:lng],
        [NSNumber numberWithFloat:level],
        [NSNumber numberWithFloat:unit],
        nil
    ];
}
@end
